File: apps/api/src/modules/facility/application/utils/cpf-checksum-fixture.test.ts
import { describe, expect, it } from "bun:test";
import fixture from "../../../../../../../packages/database/fixtures/cpf-checksum-cases.json";
import {
  isValidCpfDigits,
  normalizeLegalDocument,
} from "./facility-tax-id.utils";

/**
 * The API's half of the shared CPF fixture.
 *
 * The same rule is implemented three times — here, as `is_valid_cpf` in
 * Postgres (only the database can say which rows have a bad CPF, for the count
 * and the filtered list), and in Dart so a rep is told as they type instead of
 * after a round trip. None of the three can be removed, so the guard against
 * them drifting is that all three answer the same cases.
 *
 * `facility-tax-id.utils.test.ts` still covers this function's own edges; this
 * file exists to keep it in step with the other two.
 */
const { cases } = fixture;

describe("isValidCpfDigits against the shared fixture", () => {
  it("has cases on both sides of the verdict", () => {
    // A fixture that drifted to all-invalid would pass against a function that
    // simply returned false, and nobody would notice.
    expect(cases.some((entry) => entry.valid)).toBe(true);
    expect(cases.some((entry) => !entry.valid)).toBe(true);
  });

  for (const entry of cases) {
    it(`${entry.valid ? "accepts" : "rejects"} ${JSON.stringify(entry.raw)} — ${entry.why}`, () => {
      expect(isValidCpfDigits(entry.digits)).toBe(entry.valid);
    });
  }

  it("derives the fixture's digits from its raw value the way the app does", () => {
    // The SQL function is handed the raw string and strips punctuation itself,
    // while this validator receives already-stripped digits. If those two ever
    // disagree about what stripping means, the fixture stops comparing like
    // with like and the three implementations could diverge unnoticed.
    for (const entry of cases) {
      expect(normalizeLegalDocument(entry.raw) ?? "").toBe(entry.digits);
    }
  });
});

File: apps/mobile/test/features/explore/presentation/cpf_checksum_fixture_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:atlasmed_mobile_app/features/explore/presentation/tax_identifier.dart';
import 'package:flutter_test/flutter_test.dart';

/// The app's half of the shared CPF fixture.
///
/// Same cases as `isValidCpfDigits` in the API and `is_valid_cpf` in Postgres.
/// Three implementations exist because each answers a question the others
/// cannot — the database filters rows for the Desempenho count and its list,
/// the server is the authority for any client, and this one runs while the rep
/// is still typing. What keeps them honest is answering the same cases.
///
/// Read from the repository rather than copied: a copy would drift the moment
/// somebody added a case on one side only, which is the whole failure this
/// guards against.
void main() {
  final file = File('../../packages/database/fixtures/cpf-checksum-cases.json');
  final cases =
      (jsonDecode(file.readAsStringSync()) as Map<String, dynamic>)['cases']
          as List<dynamic>;

  test('the fixture is where this test expects it', () {
    // A moved or renamed fixture must fail loudly here rather than leave this
    // suite silently asserting nothing.
    expect(file.existsSync(), isTrue);
    expect(cases, isNotEmpty);
  });

  test('the fixture still has cases on both sides of the verdict', () {
    // Drifted to all-invalid, it would pass against a validator that always
    // returned false.
    expect(cases.any((c) => (c as Map)['valid'] == true), isTrue);
    expect(cases.any((c) => (c as Map)['valid'] == false), isTrue);
  });

  for (final entry in cases) {
    final testCase = entry as Map<String, dynamic>;
    final raw = testCase['raw'] as String;
    final expected = testCase['valid'] as bool;
    final why = testCase['why'] as String;

    test('${expected ? 'accepts' : 'rejects'} ${jsonEncode(raw)} — $why', () {
      expect(isValidCpf(raw), expected);
    });
  }

  group('cpfIssueFor', () {
    test('reports a missing CPF only for CPF clinics', () {
      expect(
        cpfIssueFor(legalDocumentType: 'CPF', legalDocument: null),
        CpfIssue.missing,
      );
      expect(
        cpfIssueFor(legalDocumentType: 'CPF', legalDocument: '   '),
        CpfIssue.missing,
      );
      // A CNPJ clinic with no CNPJ is a real problem, but not this warning's.
      expect(
        cpfIssueFor(legalDocumentType: 'CNPJ', legalDocument: null),
        isNull,
      );
    });

    test('separates an invalid CPF from a missing one', () {
      expect(
        cpfIssueFor(legalDocumentType: 'CPF', legalDocument: '529.982.247-24'),
        CpfIssue.invalid,
      );
      expect(
        cpfIssueFor(legalDocumentType: 'CPF', legalDocument: '529.982.247-25'),
        isNull,
      );
    });

    test('says nothing about a CNPJ clinic whatever its document', () {
      // is_valid_cpf is false for any 14-digit number, so a check that forgot
      // the type would flag every CNPJ clinic in the app.
      expect(
        cpfIssueFor(
          legalDocumentType: 'CNPJ',
          legalDocument: '12.345.678/0001-95',
        ),
        isNull,
      );
    });
  });
}
